### PR TITLE
feat: add endpoint for generating the authorization code

### PR DESCRIPTION
**Pull request by lcpojr**
Starts generating authorization codes.
Next step will be create the sign in and authorization page for the user to be redirected before executing this flow.

### CODE REVIEW

File: apps/authorizer/lib/input.ex [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+      def cast_to_map(%{__struct__: __MODULE__} = input) do
+        input
+        |> Map.from_struct()
+        |> Map.new(&do_cast_to_map/1)
+      end
+
+      defp do_cast_to_map(%Date{} = value), do: value
+      defp do_cast_to_map(%DateTime{} = value), do: value
+      defp do_cast_to_map(%NaiveDateTime{} = value), do: value
+      defp do_cast_to_map(%Time{} = value), do: value
```
Comment (yashin5) [inline apps/authorizer/lib/input.ex:48]:
```suggestion
      Enum.each([Date, DateTime, NaiveDateTime, Time], fn struct ->
            defp do_cast_to_map(%unquote(struct){} = value), do: value
      end)
```

File: apps/authorizer/lib/rules/commands/authorization_code_sign_in.ex [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+  end
+
+  defp build_scope(user, application, scopes) do
+    user_scopes = Enum.map(user.scopes, & &1.name)
+    app_scopes = Enum.map(application.scopes, & &1.name)
+
+    scopes
+    |> Scopes.convert_to_list()
+    |> Enum.filter(&(&1 in app_scopes))
+    |> Enum.filter(&(&1 in user_scopes))
```
Comment (yashin5) [inline apps/authorizer/lib/rules/commands/authorization_code_sign_in.ex:95]:
```suggestion
    |> Enum.filter(&(&1 in app_scopes or &1 in user_scopes))
```

File: apps/authorizer/lib/rules/commands/authorization_code_sign_in.ex [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+      "typ" => "Bearer",
+      "identity" => "user",
+      "redirect_uri" => application.redirect_uri,
+      "scope" => build_scope(user, application, scope)
+    })
+  end
+
+  defp build_scope(user, application, scopes) do
+    user_scopes = Enum.map(user.scopes, & &1.name)
+    app_scopes = Enum.map(application.scopes, & &1.name)
```
Comment (yashin5) [inline apps/authorizer/lib/rules/commands/authorization_code_sign_in.ex:90]:
```suggestion
  defp build_scope(%{scopes: user_scopes}, %{scopes: application_scopes}, scopes) do
    user_scopes = Enum.map(user.scopes, & &1.name)
    app_scopes = Enum.map(application.scopes, & &1.name)
```

File: apps/authorizer/lib/rules/commands/authorization_code_sign_in.ex [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    - The user has to be active;
+  """
+  @spec execute(input :: Input.t(), user_id :: String.t()) ::
+          {:ok, success_map()} | {:error, :unauthorized}
+  def execute(%Input{client_id: client_id} = input, user_id) when is_binary(user_id) do
+    with {:authorized?, true} <- {:authorized?, input.authorized},
+         {:app, {:ok, app}} <- {:app, Port.get_identity(%{client_id: client_id})},
+         {:flow_enabled?, true} <- {:flow_enabled?, "authorization_code" in app.grant_flows},
+         {:app_active?, true} <- {:app_active?, app.status == "active"},
+         {:same_redirect?, true} <- {:same_redirect?, app.redirect_uri == input.redirect_uri},
```
Comment (yashin5) [inline apps/authorizer/lib/rules/commands/authorization_code_sign_in.ex:39]:
```suggestion
  def execute(%Input{client_id: client_id, authorized: authorized, redirect_uri: redirect_uri} = input, user_id) when is_binary(user_id) do
    with {:authorized?, true} <- {:authorized?, authorized},
         {:app, {:ok, app}} <- {:app, Port.get_identity(%{client_id: client_id})},
         {:flow_enabled?, true} <- {:flow_enabled?, "authorization_code" in app.grant_flows},
         {:app_active?, true} <- {:app_active?, app.status == "active"},
         {:same_redirect?, true} <- {:same_redirect?, app.redirect_uri == redirect_uri},
```

File: apps/authorizer/lib/rules/commands/authorization_code_sign_in.ex [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+      {:user, {:error, :not_found}} ->
+        Logger.info("User #{user_id} not found")
+        {:error, :unauthorized}
+
+      {:user_active?, false} ->
+        Logger.info("User #{user_id} is not active")
+        {:error, :unauthorized}
+    end
+  end
```
Comment (yashin5) [inline apps/authorizer/lib/rules/commands/authorization_code_sign_in.ex:74]:
```suggestion
  def execute(
        %Input{
          client_id: client_id,
          scope: scope,
          authorized: authorized,
          redirect_uri: redirect_uri
        } = input,
        user_id
      )
      when is_binary(user_id) do
    with {:authorized?, true} <- {:authorized?, authorized},
         {:app,
          {:ok, %{status: "active", grant_flows: grant_flows, redirect_uri: ^redirect_uri} = app}} <-
           {:app, Port.get_identity(%{client_id: client_id})},
         {:flow_enabled?, true} <- {:flow_enabled?, "authorization_code" in grant_flows},
         {:user, {:ok, %{status: "active"} = user}} <-
           {:user, Port.get_identity(%GetUser{id: user_id})} do
      user
      |> generate_authorization_code(app, scope)
      |> parse_response()
    else
      {:authorized?, false} ->
        Logger.info("Client not authorized by the user")
        {:error, :unauthorized}

      {:app, %{status: _}} ->
        Logger.info("Client application #{client_id} is not active")
        {:error, :unauthorized}

      {:app, %{redirect_uri: _}} ->
        Logger.info("Request redirect_uri is different from the configured for the client")
        {:error, :unauthorized}

      {:app, {:error, :not_found}} ->
        Logger.info("Client application #{client_id} not found")
        {:error, :unauthorized}

      {:flow_enabled?, false} ->
        Logger.info("Client application #{client_id} resource_owner flow not enabled")
        {:error, :unauthorized}

      {:user, %{status: _}} ->
        Logger.info("User #{user_id} is not active")
        {:error, :unauthorized}

      {:user, {:error, :not_found}} ->
        Logger.info("User #{user_id} not found")
        {:error, :unauthorized}
    end
  end
```

File: apps/authorizer/lib/rules/commands/inputs/authorization_code_sign_in.ex [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    field :redirect_uri, :string
+    field :scope, :string
+    field :state, :string
+    field :authorized, :boolean
+  end
+
+  @doc false
+  def changeset(params) when is_map(params) do
+    %__MODULE__{}
+    |> cast(params, @required ++ @optional)
```
Comment (yashin5) [inline apps/authorizer/lib/rules/commands/inputs/authorization_code_sign_in.ex:32]: is better concat the small list in the head of the big list
```suggestion
    |> cast(params, @optional ++ @required)
```

File: apps/authorizer/test/rules/commands/authorization_code_sign_in_test.exs [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+                "iss" => "WatcherEx",
+                "jti" => _,
+                "nbf" => _,
+                "scope" => ^scope,
+                "identity" => "user",
+                "redirect_uri" => ^redirect_uri,
+                "sub" => ^subject_id,
+                "typ" => ^typ
+              }} = AuthorizationCode.verify_and_validate(authorization_code)
+    end
```
Comment (yashin5) [inline apps/authorizer/test/rules/commands/authorization_code_sign_in_test.exs:66]:
```suggestion
    test "succeeds and generates the authorization code" do
      scopes = RF.insert_list!(:scope, 3)
      %{id: user_id} = user = RF.insert!(:user)

      %{client_id: app_client_id, name: app_name, redirect_uri: app_redirect_uri} =
        app =
        RF.insert!(:client_application,
          grant_flows: ["authorization_code"],
          redirect_uri: "https://redirect-test.com"
        )

      subject_id = user_id
      client_id = app_client_id
      client_name = app_name
      redirect_uri = app_redirect_uri
      scope = scopes |> Enum.map(& &1.name) |> Enum.join(" ")

      input = %Input{
        client_id: app_client_id,
        response_type: "code",
        redirect_uri: app_redirect_uri,
        scope: scope,
        authorized: true
      }

      expect(ResourceManagerMock, :get_identity, fn %{client_id: ^app_client_id} ->
        assert app_client_id == client_id
        {:ok, %{app | public_key: nil, scopes: scopes}}
      end)

      expect(ResourceManagerMock, :get_identity, fn %{id: ^user_id} ->
        {:ok, %{user | scopes: scopes}}
      end)

      assert {:ok,
              %{
                authorization_code: authorization_code,
                expires_in: 7200,
                token_type: typ
              }} = Command.execute(input, user_id)

      assert {:ok,
              %{
                "aud" => ^client_id,
                "azp" => ^client_name,
                "exp" => _,
                "iat" => _,
                "iss" => "WatcherEx",
                "jti" => _,
                "nbf" => _,
                "scope" => ^scope,
                "identity" => "user",
                "redirect_uri" => ^redirect_uri,
                "sub" => ^subject_id,
                "typ" => ^typ
              }} = AuthorizationCode.verify_and_validate(authorization_code)
    end
  end
```

File: apps/authorizer/test/rules/commands/authorization_code_sign_in_test.exs [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+
+    test "fails if input not authorized by the user" do
+      scopes = RF.insert_list!(:scope, 3)
+      user = RF.insert!(:user)
+
+      app =
+        RF.insert!(:client_application,
+          grant_flows: ["authorization_code"],
+          redirect_uri: "https://redirect-test.com"
+        )
```
Comment (yashin5) [inline apps/authorizer/test/rules/commands/authorization_code_sign_in_test.exs:76]: same

File: apps/rest_api/lib/controllers/public/auth.ex [hunk 9]
```diff
@@ -64,19 +66,41 @@ defmodule RestAPI.Controllers.Public.Auth do
...
+          external: "#{input.redirect_uri}?code=#{resp.authorization_code}&state=#{input.state}"
+        )
+      else
+        conn
+        |> put_view(Auth)
+        |> put_status(200)
+        |> render("authorize.json", response: resp, state: input.state)
+      end
+    end
+  end
```
Comment (yashin5) [inline apps/rest_api/lib/controllers/public/auth.ex:97]:
```suggestion
    def authorize(%{private: %{session: session}} = conn, params) do
      with {:ok, %{redirect_uri: redirect_uri, state: state} = input} <-
             AuthorizationCodeSignIn.cast_and_apply(params),
           {:ok, resp} <-
             Authorizer.authorize_authorization_code_sign_in(input, session.subject_id) do
        do_authorize(redirect_uri, state, resp)
      end
    end

    defp do_authorize(nil, state, resp) do
      conn
      |> put_view(Auth)
      |> put_status(200)
      |> render("authorize.json", response: resp, state: state)
    end

    defp do_authorize(redirect_uri, state, %{authorization_code: authorization_code}) do
      redirect(conn,
        external: "#{redirect_uri}?code=#{resp.authorization_code}&state=#{state}"
      )
    end
  end
```

File: apps/authorizer/lib/rules/commands/authorization_code_sign_in.ex (hunk 5 above)
Comment (lcpojr) [reply apps/authorizer/lib/rules/commands/authorization_code_sign_in.ex:74]: I don't think it's necessary just to take out every value in another variable.
I think it's clean as it is.

File: apps/authorizer/lib/rules/commands/authorization_code_sign_in.ex (hunk 3 above)
Comment (lcpojr) [reply apps/authorizer/lib/rules/commands/authorization_code_sign_in.ex:90]: I don't think it's necessary when the usage is simple as it's here.

File: apps/authorizer/lib/rules/commands/inputs/authorization_code_sign_in.ex (hunk 6 above)
Comment (lcpojr) [reply apps/authorizer/lib/rules/commands/inputs/authorization_code_sign_in.ex:32]: Fair point, can i do it in another pull request and update all schemas?

File: apps/rest_api/lib/controllers/public/auth.ex (hunk 9 above)
Comment (lcpojr) [reply apps/rest_api/lib/controllers/public/auth.ex:97]: I really don't think it's necessary to break every struct in matches and extract the value into variables only for future use when it's simple as it is.

**Pull request merged**